File: baml_language/sdk_tests/crates/rust/type_shapes/customizable/roundtrip_tests/test_symbol_collisions.rs
//! Roundtrip coverage for the symbol-collision suite — three distinct
//! `Bar` classes at different namespace depths plus the consumers
//! (`Ipsum`, `Deep`) that compose all three.

use baml_sdk::symbol_collisions::a::b::c::d::{make_deep, round_trip_deep};
use baml_sdk::symbol_collisions::fizz::buzz::foo::{
    make_fizz_buzz_foo_bar, round_trip_fizz_buzz_foo_bar,
};
use baml_sdk::symbol_collisions::fizz::foo::{make_fizz_foo_bar, round_trip_fizz_foo_bar};
use baml_sdk::symbol_collisions::foo::{make_foo_bar, round_trip_foo_bar};
use baml_sdk::symbol_collisions::lorem::{make_ipsum, round_trip_ipsum};

#[test]
fn test_symbol_collisions_round_trip_foo_bar() {
    let bar = make_foo_bar("hi".to_string(), 2).unwrap();
    assert_eq!(round_trip_foo_bar(bar.clone()).unwrap(), bar);
}

#[test]
fn test_symbol_collisions_round_trip_fizz_foo_bar() {
    let bar = make_fizz_foo_bar("t".to_string(), 1.5).unwrap();
    assert_eq!(round_trip_fizz_foo_bar(bar.clone()).unwrap(), bar);
}

#[test]
fn test_symbol_collisions_round_trip_fizz_buzz_foo_bar() {
    let bar = make_fizz_buzz_foo_bar("f".to_string(), 2.5, true).unwrap();
    assert_eq!(round_trip_fizz_buzz_foo_bar(bar.clone()).unwrap(), bar);
}

#[test]
fn test_symbol_collisions_round_trip_ipsum() {
    let ipsum = make_ipsum(
        make_foo_bar("a".to_string(), 1).unwrap(),
        make_fizz_foo_bar("b".to_string(), 2.0).unwrap(),
        make_fizz_buzz_foo_bar("c".to_string(), 3.0, false).unwrap(),
    )
    .unwrap();
    assert_eq!(round_trip_ipsum(ipsum.clone()).unwrap(), ipsum);
}

#[test]
fn test_symbol_collisions_round_trip_deep() {
    let ipsum = make_ipsum(
        make_foo_bar("a".to_string(), 1).unwrap(),
        make_fizz_foo_bar("b".to_string(), 2.0).unwrap(),
        make_fizz_buzz_foo_bar("c".to_string(), 3.0, false).unwrap(),
    )
    .unwrap();
    let deep = make_deep(
        make_foo_bar("h".to_string(), 9).unwrap(),
        make_fizz_foo_bar("th".to_string(), 4.0).unwrap(),
        make_fizz_buzz_foo_bar("fu".to_string(), 5.0, true).unwrap(),
        ipsum,
    )
    .unwrap();
    assert_eq!(round_trip_deep(deep.clone()).unwrap(), deep);
}

File: baml_language/sdk_tests/crates/java/type_shapes/customizable/roundtrip_tests/TestSymbolCollisions.java
// Roundtrip coverage for the symbol-collision suite — three distinct `Bar`
// classes at different namespace depths plus the consumers (`Ipsum`,
// `Deep`) that compose all three.
//
// Port of python_pydantic2/type_shapes/customizable/roundtrip_tests/
// test_symbol_collisions.py — same test names, cases, inputs, assertions.
//
// java-port note: Python's collision is class-only (`Bar` appears in three
// namespaces) because free functions keep distinct BAML-declared names
// (`make_foo_bar`, `make_fizz_foo_bar`, `make_fizz_buzz_foo_bar`) that never
// collide, so each imports cleanly by name. In Java, free functions are
// static methods on a per-namespace holder that is *always* named `Fns`
// (per the conventions doc), so the same-name collision also applies to the
// function holders here, not just the `Bar` classes. Every symbol below is
// therefore referenced fully qualified rather than imported, to keep all
// five namespaces unambiguous side by side.
package roundtrip_tests;

import static org.junit.jupiter.api.Assertions.assertEquals;

import org.junit.jupiter.api.Test;

class TestSymbolCollisions {

    @Test
    void test_symbol_collisions_round_trip_foo_bar() {
        baml_sdk.symbol_collisions.foo.Bar bar =
                baml_sdk.symbol_collisions.foo.Fns.make_foo_bar("hi", 2L);
        assertEquals(bar, baml_sdk.symbol_collisions.foo.Fns.round_trip_foo_bar(bar));
    }

    @Test
    void test_symbol_collisions_round_trip_fizz_foo_bar() {
        baml_sdk.symbol_collisions.fizz.foo.Bar bar =
                baml_sdk.symbol_collisions.fizz.foo.Fns.make_fizz_foo_bar("t", 1.5);
        assertEquals(
                bar, baml_sdk.symbol_collisions.fizz.foo.Fns.round_trip_fizz_foo_bar(bar));
    }

    @Test
    void test_symbol_collisions_round_trip_fizz_buzz_foo_bar() {
        baml_sdk.symbol_collisions.fizz.buzz.foo.Bar bar =
                baml_sdk.symbol_collisions.fizz.buzz.foo.Fns.make_fizz_buzz_foo_bar(
                        "f", 2.5, true);
        assertEquals(
                bar,
                baml_sdk.symbol_collisions.fizz.buzz.foo.Fns.round_trip_fizz_buzz_foo_bar(bar));
    }

    @Test
    void test_symbol_collisions_round_trip_ipsum() {
        baml_sdk.symbol_collisions.lorem.Ipsum ipsum =
                baml_sdk.symbol_collisions.lorem.Fns.make_ipsum(
                        baml_sdk.symbol_collisions.foo.Fns.make_foo_bar("a", 1L),
                        baml_sdk.symbol_collisions.fizz.foo.Fns.make_fizz_foo_bar("b", 2.0),
                        baml_sdk.symbol_collisions.fizz.buzz.foo.Fns.make_fizz_buzz_foo_bar(
                                "c", 3.0, false));
        assertEquals(ipsum, baml_sdk.symbol_collisions.lorem.Fns.round_trip_ipsum(ipsum));
    }

    @Test
    void test_symbol_collisions_round_trip_deep() {
        baml_sdk.symbol_collisions.lorem.Ipsum ipsum =
                baml_sdk.symbol_collisions.lorem.Fns.make_ipsum(
                        baml_sdk.symbol_collisions.foo.Fns.make_foo_bar("a", 1L),
                        baml_sdk.symbol_collisions.fizz.foo.Fns.make_fizz_foo_bar("b", 2.0),
                        baml_sdk.symbol_collisions.fizz.buzz.foo.Fns.make_fizz_buzz_foo_bar(
                                "c", 3.0, false));
        baml_sdk.symbol_collisions.a.b.c.d.Deep deep =
                baml_sdk.symbol_collisions.a.b.c.d.Fns.make_deep(
                        baml_sdk.symbol_collisions.foo.Fns.make_foo_bar("h", 9L),
                        baml_sdk.symbol_collisions.fizz.foo.Fns.make_fizz_foo_bar("th", 4.0),
                        baml_sdk.symbol_collisions.fizz.buzz.foo.Fns.make_fizz_buzz_foo_bar(
                                "fu", 5.0, true),
                        ipsum);
        assertEquals(deep, baml_sdk.symbol_collisions.a.b.c.d.Fns.round_trip_deep(deep));
    }
}
